File: test/detection/compare_results.py
import pytest
import json
import os

from typing import Union

CONF_ERROR_MARGIN = 0.1 # +-0.01 difference allowed in confidence scores
BBOX_ERROR_MARGIN = 50  # +-50 pixels difference allowed in bounding box coordinates


@pytest.fixture
def results_ultralytics() -> Union[dict, None]:

   results_ultralytics_path = "results/results_ultralytics.json"

   if not os.path.exists(results_ultralytics_path):
      pytest.skip(f"{results_ultralytics_path} does not exist")
      return None
   
   with open(results_ultralytics_path, "r") as f:
      return json.load(f)

@pytest.fixture
def results_cpp() -> Union[dict, None]:
   results_cpp_path = "results/results_cpp.json"

   if not os.path.exists(results_cpp_path):
      pytest.skip(f"{results_cpp_path} does not exist")
      return None
   
   with open(results_cpp_path, "r") as f:
      return json.load(f)

def test_results_not_empty(results_ultralytics : Union[dict, None], results_cpp : Union[dict, None]):
    
   if results_ultralytics is None:
      pytest.skip("results_ultralytics is None")

   if results_cpp is None:
      pytest.skip("results_cpp is None")

   if len(results_ultralytics) == 0:
      pytest.skip("results_ultralytics is empty")

   if len(results_cpp) == 0:
      pytest.skip("results_cpp is empty")
   
def test_compare_models_names(results_ultralytics : dict, results_cpp : dict):
    
   models_names_ultralytics = set(results_ultralytics.keys())
   models_names_cpp = set(results_cpp.keys())

   for model_name in models_names_ultralytics:
      assert model_name in models_names_cpp, f"Model {model_name} is missing in results_cpp"

def test_compare_weights_paths(results_ultralytics : dict, results_cpp : dict):
    
   for model_name in results_ultralytics.keys():

      weights_path_ultralytics = results_ultralytics[model_name].get("weights_path", "")
      weights_path_cpp = results_cpp[model_name].get("weights_path", "")

      assert weights_path_ultralytics == weights_path_cpp, f"Weights path mismatch for model {model_name}: {weights_path_ultralytics} != {weights_path_cpp}"

def test_compare_images_counts(results_ultralytics : dict, results_cpp : dict):
    
   for model_name in results_ultralytics.keys():

      results_ultralytics_model = results_ultralytics[model_name]["results"]
      results_cpp_model = results_cpp[model_name]["results"]

      assert len(results_ultralytics_model) == len(results_cpp_model), f"Number of results mismatch for model {model_name}: {len(results_ultralytics_model)} != {len(results_cpp_model)}"

def test_compare_images_paths(results_ultralytics : dict, results_cpp : dict):
    
   for model_name in results_ultralytics.keys():

      results_ultralytics_model = results_ultralytics[model_name]["results"]
      results_cpp_model = results_cpp[model_name]["results"]

      for i in range(len(results_ultralytics_model)):

         image_path_ultralytics = results_ultralytics_model[i].get("image_path")
         image_path_cpp = results_cpp_model[i].get("image_path")

         assert image_path_ultralytics == image_path_cpp, f"Image path mismatch for model {model_name}, image {i}: {image_path_ultralytics} != {image_path_cpp}"
   

def test_compare_detections_count(results_ultralytics : dict, results_cpp : dict):

   for model_name in results_ultralytics.keys():

      results_ultralytics_model = results_ultralytics[model_name]["results"]
      results_cpp_model = results_cpp[model_name]["results"]

      for i in range(len(results_ultralytics_model)):

         detections_ultralytics = results_ultralytics_model[i].get("inference_results", [])
         detections_cpp = results_cpp_model[i].get("inference_results", [])

         image_path = results_ultralytics_model[i].get("image_path")

         assert len(detections_ultralytics) == len(detections_cpp), f"Number of detections mismatch for model {model_name}, image :  {image_path}: ultralytics: {len(detections_ultralytics)} != cpp: {len(detections_cpp)}"

def test_compare_detections(results_ultralytics : dict, results_cpp : dict):

   for model_name in results_ultralytics.keys():

      results_ultralytics_model = results_ultralytics[model_name]["results"]
      results_cpp_model = results_cpp[model_name]["results"]

      for i in range(len(results_ultralytics_model)):

         detections_ultralytics = results_ultralytics_model[i].get("inference_results", [])
         detections_cpp = results_cpp_model[i].get("inference_results", [])

         image_path = results_ultralytics_model[i].get("image_path")

         for j in range(len(detections_ultralytics)):

            detection_ultralytics = detections_ultralytics[j]

            class_id_ultralytics = detection_ultralytics.get("class_id")
            conf_ultralytics = detection_ultralytics.get("confidence")
            bbox_ultralytics = detection_ultralytics.get("bbox") 

            for k in range(len(detections_cpp)):

               is_class_found = False

               detection_cpp = detections_cpp[k]

               class_id_cpp = detection_cpp.get("class_id")

               if class_id_ultralytics == class_id_cpp:

                  bbox_cpp = detection_cpp.get("bbox")

                  left_diff = abs(bbox_ultralytics["left"] - bbox_cpp["left"])
                  top_diff = abs(bbox_ultralytics["top"] - bbox_cpp["top"])
                  width_diff = abs(bbox_ultralytics["width"] - bbox_cpp["width"])
                  height_diff = abs(bbox_ultralytics["height"] - bbox_cpp["height"])

                  print(f"""
                  Model: {model_name}, Image: {image_path}, Class ID: {class_id_ultralytics}
                  BBox Ultralytics: {bbox_ultralytics}
                  BBox Cpp: {bbox_cpp}
                  Diffs - Left: {left_diff}, Top: {top_diff}, width: {width_diff}, height: {height_diff}
                  """)

                  if (left_diff <= BBOX_ERROR_MARGIN and
                      top_diff <= BBOX_ERROR_MARGIN and
                      width_diff <= BBOX_ERROR_MARGIN and
                      height_diff <= BBOX_ERROR_MARGIN):

                     conf_cpp = detection_cpp.get("confidence")

                     conf_diff = abs(conf_ultralytics - conf_cpp)

                     print(f"""
                     Model: {model_name}, Image: {image_path}, Class ID: {class_id_ultralytics}
                     Ultralytics Conf: {conf_ultralytics}, Cpp Conf: {conf_cpp}
                     Diff: {conf_diff}
                     """)

                     assert conf_diff <= CONF_ERROR_MARGIN, f"Confidence mismatch for model {model_name}, image :  {image_path}, class_id: {class_id_ultralytics}: ultralytics: {conf_ultralytics} != cpp: {conf_cpp}"

                     is_class_found = True
                     break

            
            assert is_class_found, f"Class ID {class_id_ultralytics} not found in cpp results for model {model_name}, image: {image_path}"

            

File: test/detection/compare_results.cpp
#include <gtest/gtest.h>
#include <fstream>
#include <string>
#include <map>
#include <vector>
#include <cmath>
#include <nlohmann/json.hpp>

#define STRING(x) #x
#define XSTRING(x) STRING(x)

using json = nlohmann::json;

constexpr double CONF_ERROR_MARGIN = 0.1; // +-0.1 difference allowed in confidence scores
constexpr int BBOX_ERROR_MARGIN = 50;     // +-50 pixels difference allowed in bounding box coordinates

json read_json(const std::string& path) {
    std::ifstream f(path);
    if (!f.is_open()) {
        throw std::runtime_error("File not found: " + path);
    }
    json j;
    f >> j;
    return j;
}

class ResultsFixture : public ::testing::Test {
protected:
    json results_ultralytics;
    json results_cpp;
    std::string basePath = XSTRING(BASE_PATH_DETECTION); // Base path defined in CMakeLists.txt


    void SetUp() override {
        ASSERT_NO_THROW(results_ultralytics = read_json(basePath + "results/results_ultralytics.json"));
        ASSERT_NO_THROW(results_cpp = read_json(basePath + "results/results_cpp.json"));
    }
};

TEST_F(ResultsFixture, ResultsNotEmpty) {
    ASSERT_FALSE(results_ultralytics.empty()) << "results_ultralytics is empty";
    ASSERT_FALSE(results_cpp.empty()) << "results_cpp is empty";
}

TEST_F(ResultsFixture, CompareModelsNames) {

    std::set<std::string> models_ultra, models_cpp;

    for (auto& el : results_ultralytics.items()) models_ultra.insert(el.key());
    for (auto& el : results_cpp.items()) models_cpp.insert(el.key());

    for (const auto& name : models_ultra) {
        ASSERT_TRUE(models_cpp.count(name)) << "Model " << name << " is missing in results_cpp";
    }
}

TEST_F(ResultsFixture, CompareWeightsPaths) {

    for (auto& el : results_ultralytics.items()) {

        const std::string& model_name = el.key();

        std::string weights_ultra = el.value().value("weights_path", "");
        std::string weights_cpp = results_cpp[model_name].value("weights_path", "");

        ASSERT_EQ(weights_ultra, weights_cpp) << "Weights path mismatch for model " << model_name;
    }
}

TEST_F(ResultsFixture, CompareImagesCounts) {

    for (auto& el : results_ultralytics.items()) {

        const std::string& model_name = el.key();

        auto& ultra_results = el.value()["results"];
        auto& cpp_results = results_cpp[model_name]["results"];

        ASSERT_EQ(ultra_results.size(), cpp_results.size())
            << "Number of results mismatch for model " << model_name;
    }
}

TEST_F(ResultsFixture, CompareImagesPaths) {

    for (auto& el : results_ultralytics.items()) {

        const std::string& model_name = el.key();

        auto& ultra_results = el.value()["results"];
        auto& cpp_results = results_cpp[model_name]["results"];

        for (size_t i = 0; i < ultra_results.size(); ++i) {

            std::string path_ultra = ultra_results[i].value("image_path", "");
            
            bool path_found = false;
            for (size_t j = 0; j < cpp_results.size(); ++j) {
                std::string path_cpp = cpp_results[j].value("image_path", "");
                if (path_ultra == path_cpp) {
                    path_found = true;
                    break;
                }
            }

            ASSERT_TRUE(path_found)
                << "Image path mismatch for model " << model_name << ", image " << i;
        }
    }
}

TEST_F(ResultsFixture, CompareDetectionsCount) {

    for (auto& el : results_ultralytics.items()) {

        const std::string& model_name = el.key();

        auto& ultra_results = el.value()["results"];
        auto& cpp_results = results_cpp[model_name]["results"];

        for (size_t i = 0; i < ultra_results.size(); ++i) {

            std::string image_path = ultra_results[i].value("image_path", "");

            auto detections_ultra = ultra_results[i].value("inference_results", json::array());

            for (size_t j = 0; j < cpp_results.size(); ++j) {
                if (cpp_results[j].value("image_path", "") == image_path) {
                    auto detections_cpp = cpp_results[j].value("inference_results", json::array());
                    ASSERT_EQ(detections_cpp.size(), detections_ultra.size())
                        << "Number of detections mismatch for model " << model_name << ", image: " << image_path;
                    break;
                }
            }
            
        }
    }
}

TEST_F(ResultsFixture, CompareDetections) {

    for (auto& el : results_ultralytics.items()) {

        const std::string& model_name = el.key();

        auto& ultra_results = el.value()["results"];
        auto& cpp_results = results_cpp[model_name]["results"];

        for (size_t i = 0; i < ultra_results.size(); ++i) {

            auto detections_ultra = ultra_results[i].value("inference_results", json::array());

            std::string image_path = ultra_results[i].value("image_path", "");

            for( size_t j = 0; j < cpp_results.size(); ++j) {

                if (cpp_results[j].value("image_path", "") == image_path) {

                    auto detections_cpp = cpp_results[j].value("inference_results", json::array());
                    
                    for (size_t j = 0; j < detections_ultra.size(); ++j) {

                        auto& det_ultra = detections_ultra[j];
                        int class_id_ultra = det_ultra.value("class_id", -1);
                        double conf_ultra = det_ultra.value("confidence", 0.0);
                        auto bbox_ultra = det_ultra["bbox"];
                        bool is_class_found = false;

                        for (size_t k = 0; k < detections_cpp.size(); ++k) {

                            auto& det_cpp = detections_cpp[k];

                            int class_id_cpp = det_cpp.value("class_id", -2);

                            bool already_matched = detections_cpp[k].value("_matched", false);

                            if (!already_matched && class_id_ultra == class_id_cpp) {

                                auto bbox_cpp = det_cpp["bbox"];

                                int left_diff = std::abs(bbox_ultra["left"].get<int>() - bbox_cpp["left"].get<int>());
                                int top_diff = std::abs(bbox_ultra["top"].get<int>() - bbox_cpp["top"].get<int>());
                                int width_diff = std::abs(bbox_ultra["width"].get<int>() - bbox_cpp["width"].get<int>());
                                int height_diff = std::abs(bbox_ultra["height"].get<int>() - bbox_cpp["height"].get<int>());

                                if (left_diff <= BBOX_ERROR_MARGIN &&
                                    top_diff <= BBOX_ERROR_MARGIN &&
                                    width_diff <= BBOX_ERROR_MARGIN &&
                                    height_diff <= BBOX_ERROR_MARGIN) {

                                    double conf_cpp = det_cpp.value("confidence", 0.0);
                                    double conf_diff = std::abs(conf_ultra - conf_cpp);

                                    ASSERT_LE(conf_diff, CONF_ERROR_MARGIN)
                                        << "Confidence mismatch for model " << model_name
                                        << ", image: " << image_path << ", class_id: " << class_id_ultra
                                        << ": ultralytics: " << conf_ultra << " != cpp: " << conf_cpp;

                                    is_class_found = true;

                                    detections_cpp[k]["_matched"] = true; // Mark as matched
                                    
                                    break;
                                }
                            }
                        }
                        
                        ASSERT_TRUE(is_class_found)
                            << "Class ID " << class_id_ultra << " not found in cpp results for model "
                            << model_name << ", image: " << image_path;
                    }

                    break;
                }
            }
        }
    }
}
